### PR TITLE
SWIFT-844: ObjectID and Generator

### PR DESCRIPTION
**Pull request by nbbeeken**
https://jira.mongodb.org/browse/SWIFT-844

Fleshed out the remaining implementation of ObjectID

### CODE REVIEW

File: Sources/BSON/BSONObjectID.swift [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    }
+}
+
+internal class ObjectIDGenerator {
+    private let randomNumber: [UInt8]
+    private var counter: NIOAtomic<UInt32>
+    internal init() {
+        // 5 bytes of a random number
+        self.randomNumber = Array(withUnsafeBytes(
+            of: UInt64.random(in: 0...0xFF_FFFF_FFFF).bigEndian, [UInt8].init
```
Comment (patrickfreed) [inline Sources/BSON/BSONObjectID.swift:72]: here and elsewhere, it's probably a bit clearer to just use the `.max` constants (e.g. `UInt64.max` in this case)

File: Sources/BSON/BSONObjectID.swift [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        buffer.writeBytes(self.oid)
+    }
+}
+
+internal class ObjectIDGenerator {
+    private let randomNumber: [UInt8]
+    private var counter: NIOAtomic<UInt32>
+    internal init() {
+        // 5 bytes of a random number
+        self.randomNumber = Array(withUnsafeBytes(
```
Comment (patrickfreed) [inline Sources/BSON/BSONObjectID.swift:71]: is the second `Array` wrapping necessary? I think `withUnsafeBytes(..., [UInt8].init)` should be sufficient.

File: Sources/BSON/BSONObjectID.swift [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        // 3 byte counter started randomly per process
+        self.counter = NIOAtomic<UInt32>.makeAtomic(value: UInt32.random(in: 0..<0xFFFFFF))
+    }
+
+    internal func generate() -> [UInt8] {
+        // roll over counter
+        _ = self.counter.compareAndExchange(expected: 0xFFFFFF, desired: 0x00)
+        // fetch current timestamp
+        let timestamp = UInt32(Date().timeIntervalSince1970)
+        var buffer = [UInt8]()
```
Comment (patrickfreed) [inline Sources/BSON/BSONObjectID.swift:83]: this array can also be initialized with 12 values via the `init(repeating:length:)`.

File: Sources/BSON/BSONObjectID.swift [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        }
+        return .objectID(BSONObjectID(bytes))
+    }
+
+    internal func write(to buffer: inout ByteBuffer) {
+        buffer.writeBytes(self.oid)
+    }
+}
+
+internal class ObjectIDGenerator {
```
Comment (patrickfreed) [inline Sources/BSON/BSONObjectID.swift:66]: this type could use a docstring explaining why its here and what it does.

File: Sources/BSON/BSONObjectID.swift [hunk 5]
```diff
@@ -0,0 +1,89 @@
+import Foundation
+import NIO
+import NIOConcurrencyHelpers
+
+/// A struct to represent the BSON ObjectID type.
+public struct BSONObjectID: Equatable, Hashable, CustomStringConvertible {
+    /// This `BSONObjectID`'s data represented as a `String`.
+    public var hex: String { self.oid.reduce("") { $0 + String(format: "%02x", $1) } }
```
Comment (patrickfreed) [inline Sources/BSON/BSONObjectID.swift:8]: some tests verifying that converting an `ObjectID` to/from a hex string happens losslessly would be nice.

File: Sources/BSON/BSONObjectID.swift [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    internal init(_ bytes: [UInt8]) {
+        self.oid = bytes
+    }
+
+    /// Initializes an `BSONObjectID` from the provided hex `String`.
+    /// - Throws:
+    ///   - `BSONError.InvalidArgumentError` if string passed is not a valid BSONObjectID
+    /// - SeeAlso: https://github.com/mongodb/specifications/blob/master/source/objectid.rst
+    public init(_ hex: String) throws {
+        guard hex.utf8.count == 24 else {
```
Comment (patrickfreed) [inline Sources/BSON/BSONObjectID.swift:30]: instead of a magic number, we should define a static constant under `ObjectID` that is it's length (i.e. 12). Then we can do 2 * that constant here, and the local variable `len` can just go away.

File: Sources/BSON/BSONObjectID.swift (hunk 1 above)
Comment (nbbeeken) [reply Sources/BSON/BSONObjectID.swift:72]: I think that would reduce the randomness, not sure though. It's supposed to be a 5 byte number and if I go beyond 5 bytes for the upper limit than that upper limit has the chance to occur many times as opposed to just once, does that make sense?

File: Sources/BSON/BSONObjectID.swift (hunk 2 above)
Comment (nbbeeken) [reply Sources/BSON/BSONObjectID.swift:71]: I take a slice of it so it's an ArraySlice if I don't cast it, I can make it better using a [UInt8]() constructor

File: Sources/BSON/BSONObjectID.swift (hunk 3 above)
Comment (nbbeeken) [reply Sources/BSON/BSONObjectID.swift:83]: I'm extending the array below I can do this but how do I assign to array slices? or do I just loop and assign per index?

File: Sources/BSON/BSONObjectID.swift (hunk 1 above)
Comment (patrickfreed) [reply Sources/BSON/BSONObjectID.swift:72]: Oh I see. Honestly I just saw a bunch of F's and assumed you were typing the literal for the max value, but I see you were only going for 5 bytes instead of the full 8.

File: Sources/BSON/BSONObjectID.swift (hunk 3 above)
Comment (patrickfreed) [reply Sources/BSON/BSONObjectID.swift:83]: Oh good point. In this case it would probably be better to use [`reserveCapacity()`](https://developer.apple.com/documentation/swift/array/1538966-reservecapacity) here instead then.

File: Sources/BSON/BSONObjectID.swift [hunk 7]
```diff
@@ -26,7 +29,22 @@ public struct BSONObjectID: Equatable, Hashable, CustomStringConvertible {
     ///   - `BSONError.InvalidArgumentError` if string passed is not a valid BSONObjectID
     /// - SeeAlso: https://github.com/mongodb/specifications/blob/master/source/objectid.rst
     public init(_ hex: String) throws {
-        fatalError("ah")
+        guard hex.utf8.count == (BSONObjectID.LENGTH * 2) else {
+            throw BSONError.InvalidArgumentError(message: "Cannot create ObjectId from \(hex)")
```
Comment (kmahar) [inline Sources/BSON/BSONObjectID.swift:33]: since we know the issue here (invalid length) maybe we could say that in the error message?

File: Sources/BSON/BSONObjectID.swift [hunk 8]
```diff
@@ -26,7 +29,22 @@ public struct BSONObjectID: Equatable, Hashable, CustomStringConvertible {
...
+        }
+        let len = hex.utf8.count / 2
+        var data = [UInt8](repeating: 0, count: 12)
+        for i in 0..<len {
+            let j = hex.index(hex.startIndex, offsetBy: i * 2)
+            let k = hex.index(j, offsetBy: 2)
+            let bytes = hex[j..<k]
+            if let num = UInt8(bytes, radix: 16) {
```
Comment (kmahar) [inline Sources/BSON/BSONObjectID.swift:41]: nit: seems like a natural place to use `guard let num...` rather than if/else

File: Sources/BSON/BSONObjectID.swift [hunk 9]
```diff
@@ -26,7 +29,22 @@ public struct BSONObjectID: Equatable, Hashable, CustomStringConvertible {
...
+                data[i] = num
+            } else {
+                throw BSONError.InvalidArgumentError(message: "Cannot create UInt8 from \(bytes)")
+            }
+        }
+        self = BSONObjectID([UInt8](data))
```
Comment (kmahar) [inline Sources/BSON/BSONObjectID.swift:47]: 1. isn't it already a `[UInt8]`? is the cast necessary?
2. I can't comment on the initializer that accepts bytes since it's old code but right now we could accidentally circumvent the length requirement internally by using that. maybe consider moving the length check/error into there, or adding a precondition in that init (evaluated in debug only) that makes sure we have 12 bytes

File: Sources/BSON/BSONObjectID.swift [hunk 10]
```diff
@@ -36,7 +54,7 @@ extension BSONObjectID: BSONValue {
     internal var bson: BSON { .objectID(self) }
 
     internal static func read(from buffer: inout ByteBuffer) throws -> BSON {
-        guard let bytes = buffer.readBytes(length: 12) else {
+        guard let bytes = buffer.readBytes(length: BSONObjectID.LENGTH) else {
             throw BSONError.InternalError(message: "Cannot read 12 bytes for BSONObjectID")
```
Comment (kmahar) [inline Sources/BSON/BSONObjectID.swift:58]: should we use the length constant in the error message too?

File: Sources/BSON/ByteBuffer+BSON.swift [hunk 11]
```diff
@@ -10,14 +10,27 @@ extension ByteBuffer {
...
+    internal func getCString(at offset: Int) throws -> String {
+        let key = try getBSONKey(at: offset).dropLast()
+        guard let string = String(bytes: key, encoding: .utf8) else {
+            throw BSONError.InternalError(message: "Failed to decode BSONKey as UTF8: \(key)")
+        }
+        return string
+    }
+
+    /// Returns the C String key including the null byte, for ease of comparison and byte counting
+    internal func getBSONKey(at offset: Int) throws -> [UInt8] {
```
Comment (kmahar) [inline Sources/BSON/ByteBuffer+BSON.swift:28]: I think this name is confusing because C strings aren't just used for BSON keys. can we come up with a name that better clarifies what exactly this method does? for example, "getCStringBytes" ? I also don't see this method used outside of `getCString`, where will we be using it otherwise?

File: Tests/BSONTests/BSONCorpusTests.swift [hunk 12]
```diff
@@ -109,6 +109,9 @@ final class BSONCorpusTests: BSONTestCase {
                         XCTFail("Unable to interpret canonical_bson as Data")
                         return
                     }
+                    let description = "\(testFile.description)-\(test.description)"
```
Comment (kmahar) [inline Tests/BSONTests/BSONCorpusTests.swift:112]: this looks like debugging changes?

File: Tests/DisabledTests/BSONValueTests.swift [hunk 13]
```diff
@@ -1,11 +1,9 @@
-import CLibMongoC
+@testable import BSON
```
Comment (kmahar) [inline Tests/DisabledTests/BSONValueTests.swift:1]: given that this file is still in "disabled", does it make sense to include these changes now or no?

File: Sources/BSON/BSONObjectID.swift [hunk 14]
```diff
@@ -26,7 +29,22 @@ public struct BSONObjectID: Equatable, Hashable, CustomStringConvertible {
...
+        let len = hex.utf8.count / 2
+        var data = [UInt8](repeating: 0, count: 12)
+        for i in 0..<len {
+            let j = hex.index(hex.startIndex, offsetBy: i * 2)
+            let k = hex.index(j, offsetBy: 2)
+            let bytes = hex[j..<k]
+            if let num = UInt8(bytes, radix: 16) {
+                data[i] = num
+            } else {
+                throw BSONError.InvalidArgumentError(message: "Cannot create UInt8 from \(bytes)")
```
Comment (kmahar) [inline Sources/BSON/BSONObjectID.swift:44]: i think it would be a bit clearer to say something like "Invalid hexadecimal character x" here

File: Sources/BSON/BSONObjectID.swift [hunk 15]
```diff
@@ -46,3 +69,29 @@ extension BSONObjectID: BSONValue {
         buffer.writeBytes(self.oid)
     }
 }
+
+/// A class responsible for generating ObjectIDs for a given instance of this library
+/// An ObjectID consists of a random number for this process, a timestamp, and a counter
+internal class ObjectIDGenerator {
+    private let randomNumber: [UInt8]
+    private var counter: NIOAtomic<UInt32>
```
Comment (kmahar) [inline Sources/BSON/BSONObjectID.swift:77]: it might be useful to note somewhere, either here or below where it's used that we only use the three lower bytes of this value

File: Tests/BSONTests/BSONTests.swift [hunk 16]
```diff
@@ -71,26 +71,42 @@ func toByteString(_ bytes: [UInt8]?) -> String {
...
 }
 
 public extension Array where Element == UInt8 {
-    var byteString: String {
-        BSONTests.toByteString(self)
+    func toByteString(ascii: Bool = true) -> String {
+        BSONTests.toByteString(self, ascii: ascii)
+    }
+
+    init?(hexString: String) {
```
Comment (kmahar) [inline Tests/BSONTests/BSONTests.swift:84]: this looks very similar to our logic for creating an ObjectID from a hex string. Perhaps we need some internal helper method that we can use in both places?

File: Tests/BSONTests/BSONTests.swift [hunk 17]
```diff
@@ -53,14 +53,14 @@ public func retrieveSpecTestFiles<T: Decodable>(
         }
 }
 
-func toByteString(_ bytes: [UInt8]?) -> String {
+func toByteString(_ bytes: [UInt8]?, ascii: Bool = false) -> String {
```
Comment (kmahar) [inline Tests/BSONTests/BSONTests.swift:56]: document what the ascii option does?

File: Sources/BSON/BSON.swift [hunk 18]
```diff
@@ -211,6 +211,67 @@ extension BSON {
     }
 }
 
+extension BSON {
```
Comment (patrickfreed) [inline Sources/BSON/BSON.swift:214]: ditto this doesn't seem related to the ticket at hand here, are these needed for the `ObjectID` tests or something?

File: Tests/BSONTests/BSONObjectIDTests.swift [hunk 19]
```diff
@@ -0,0 +1,81 @@
+@testable import BSON
+import Foundation
+import Nimble
+
+extension BSONObjectID {
+    // timestamp
+    internal var timestamp: Int {
```
Comment (patrickfreed) [inline Tests/BSONTests/BSONObjectIDTests.swift:7]: instead of all this bit shifting, I think it would be clearer to just use a temporrary `ByteBuffer` or `withUnsafeBytes` + `load` approach that we used somewhere else.

**Comment by kmahar** [comment]
Just did another pass. I agree with all of Patrick's outstanding comments; I don't have any further comments myself right now though.

File: Sources/BSON/BSON.swift (hunk 18 above)
Comment (nbbeeken) [reply Sources/BSON/BSON.swift]: Yea to both comments, this was the last ticket that was carrying around 50+ commits of messy history I learned my rebase lesson on this project
Anyway cleaned it up, this just has OID changes now

File: Sources/BSON/BSONObjectID.swift [hunk 20]
```diff
@@ -1,23 +1,32 @@
...
+
     /// Initializes a new `BSONObjectID`.
     public init() {
-        fatalError("ah")
+        self.oid = Self.generator.generate()
     }
 
     /// Initializes a new `BSONObjectID`.
     internal init(_ bytes: [UInt8]) {
+        precondition(
```
Comment (patrickfreed) [inline Sources/BSON/BSONObjectID.swift:26]: this precondition will only be triggered in debug mode, which is why we generally use `fatalError`. Is that intentional here?

Comment (kmahar) [reply Sources/BSON/BSONObjectID.swift:26]: that was a suggestion by me - since this is an internal method and we only call it from places that already validate the length beforehand, it seems like overkill to check twice in release mode. but at the same time it seemed like this method should do *some* kind of validation

Comment (nbbeeken) [reply Sources/BSON/BSONObjectID.swift:26]: It is intention in that @kmahar suggested it, I think because this is internal and used currently only when reading out of the ByteBuffer, which has already asserted it has read 12 bytes

Comment (patrickfreed) [reply Sources/BSON/BSONObjectID.swift:26]: sgtm
